feat : 리그 오브 레전드 실시간 이벤트 생성 API

Ra99it opened the pull request:
리그 오브 레전드를 벤치마킹한 실시간 이벤트 생성 API입니다.

API는 다음과 같은 기능을 수행합니다.

## 구조 및 순서
- Main.java 
  - Room.java
    - Play.java 
    
## API 설명

1. Player/10 의 Room을 생성합니다. Player는 조정이 가능합니다.
2. Room을 생성하고, 각 Room당 10명의 Player 병렬 스레드를 실행합니다.
3. 생성된 Room마다  각기 **다른 durationSeconds(실행 시간)** 을 가지고 있습니다.
4. durationSeconds 동안 Player 스레드는 다음과 같은 정보를 실시간으로 생성합니다.
5. RoomID, id, createdRoomDate, ip, account, champion, datetime, x, y, inputkey, status, deathCount, ingameTime
6. 생성된 이벤트는 Kafka Producer로 전송됩니다.
7. Player 100명을 기준으로 시간 당 23,430개의 데이터를 생성합니다

> Room마다 다른 durationSeconds을 가지는 이유는 생동감 있는 환경을 연출하기 위함입니다.

---

Ra99it committed:
Add feature : LoL Log Genertor API

Ra99it committed:
Add feature: LoL log Genertor API

Ra99it committed:
Merge pull request #2 from f-lab-edu/feature/#1

#1 리그 오브 레전드 실시간 이벤트 생성 API

Ra99it committed:
refactor : LoL PlyerEventGenertor -> Lol PlayerEventGenertor 오타 수정

Ra99it committed:
chore : LoL_PlayerEventGenertor 중복 이슈로 삭제합니다.

Ra99it committed:
chore : LoL_plyerEventGenertor -> Lol_PlayerEventGenertor 파일명 변경

Ra99it committed:
Merge branch 'develop' of https://github.com/f-lab-edu/league-of-legends-data-solution into develop

Ra99it committed:
chore : LoL_PlyerEventGenertor 오타로 인한 삭제

Ra99it committed:
feat : Docker Swarm 환경 Kafka Cluster 추가

Ra99it committed:
feat : Docker Swarm 환경 Trino Cluster 구축

Ra99it committed:
feat : Docker Swarm 환경 Airflow 환경 구축

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Main.java:
```diff
@@ -0,0 +1,39 @@
+
+        try {
+            latch.await();
+        }catch (InterruptedException e) {
```
띄어쓰기를 보니 전체적으로 코드 스타일이 자동으로 적용되어 있지 않은 것 같습니다.
여러 사람들과 협업 시에는 일관된 코드 작성을 위해 가장 먼저 코드 컨벤션을 정의합니다.
IntelliJ를 사용하신다면 IDE에서 google style guide 적용하고
Actions on Save를 통해 자동 정렬하도록 설정하는게 좋을 것 같습니다.

Ra99it replied:
원활한 협업을 위해 코드 규칙을 정의하는 것에 대해서 자세히 알게 되었고  google style guide로 적용을 완료했습니다.

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Main.java:
```diff
@@ -0,0 +1,39 @@
+
+public class Main {
+    static int userNum = 100 ; // 플레이어의 수를 지정합니다.
+    static int roomNum = userNum/10; // 방의 수를 변수로 저장합니다.
```
플레이어와 방의 수는 항상 고정이어야 할까요?
데이터 생성 기능임을 생각했을 때 이 부분은 외부 설정으로 분리하고
실행할 때 값을 주입하는 방식으로 수정하는 방법은 어떨까요?

Ra99it replied:
맞습니다. 항상 IDE에서 소스를 실행시켰기 때문에 이런 부분은 크게 생각하지 않았던 거 같습니다. 현재 플레이어의 수를 외부에서 입력하게끔 방식을 변경하였습니다.

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Play.java:
```diff
@@ -0,0 +1,240 @@
+
+    @Override
+    public void run() {
+        System.out.println(account+ "님이 참가하셨습니다. 챔피언: "+ champion  + " ( ip: "  + ipAddr +", account="+ account +", sessionRoomID=" + sessionRoomID + ", durationSeconds=" + durationSeconds);
```
로그 확장성을 위해 sysout으로 작성하기보다 logger를 활용하는게 어떨까요?
각 언어마다 다양한 라이브러리를 지원하고 있으니 검토해보는 것도 좋을 것 같습니다.
log4j는 과거에 보안 취약점 이슈가 있었으니 최신 버전 사용하는 것을 권장드려요.

Ra99it replied:
해당 부분은 플레이어 수에 맞게 참가 했는지 확인하기 위해 작성된 코드입니다. 실제로 출력만 진행하고 데이터를 수집하지는 않습니다. 이러한 상황에도 looger를 사용하는 것이 좋을까요?

f-lab-johnp replied:
네 그렇다면 이번에는 적용하지 않으셔도 괜찮습니다!

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Main.java:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    public static String getSessionRoomID() { // 생성된 Room의 ID를 생성하는 함수입니다.
```
이 부분도 코드 컨벤션을 정의할 때 다시 검토해보는게 좋을 것 같습니다.
일반적으로 함수의 주석을 작성할 때에는 javadoc 이라는 표준 형식을 사용합니다.

Ra99it replied:
적극적으로 반영하겠습니다

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Play.java:
```diff
@@ -0,0 +1,240 @@
+        //props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "kafka-cluster-01:9092,kafka-cluster-02:9092,kafka-cluster-03:9092");
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "SoloGameDataGenerator");
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
```
위의 `ProducerConfig` 부분도 외부 설정을 통해 주입하는게 어떨까요?
특히 kafka cluster 호스트 정보와 같은 민감한 정보는 보안 문제로 코드에 값을 노출하는 것을 권장하지 않습니다.

Ra99it replied:
보안을 위해서 IP 대신 Hostname으로 작성하였는데 이 부분도 보안과 유연성 부분의 문제를 있다는 걸 인지하였습니다. 

혹시 SERIALIZER와 같은 정보도 민감한 정보일까요? 현재 호스트 정보는 외부에서 입력하게끔 변경하였습니다.

f-lab-johnp replied:
`SERIALIZER`는 그대로 사용하셔도 괜찮을 것 같습니다. 수고하셨습니다!

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Main.java:
```diff
@@ -0,0 +1,39 @@
+            executor.execute(new Room(sessionRoomID, createRoomDate, userNum));
+        });
+
+        executor.shutdown();
```
이 부분은 모든 작업이 완료된 이후에 쓰레드 풀을 종료해야하지 않을까요?
finally block으로 이동하는건 어떨지 제안드려봅니다.

Ra99it replied:
완료했습니다.

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Room.java:
```diff
@@ -0,0 +1,129 @@
+                }else {
+                    championName = "zilean";
+                }
+                return championName;
```
이 부분은 가독성, 확장성을 위해 Enum Class로 분리하는건 어떨까요?
참고: https://techblog.woowahan.com/2527/

Ra99it replied:
소스가 번잡하다는 걸 인지하였고 Enum Class로 간결하게 작성을 완료했습니다.

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Play.java:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    private int getX_viktor(){
```
java에서 메서드 이름은 CamelCase를 따르는 것이 표준 컨벤션입니다.
위 메서드는 SnakeCase를 혼합하여 사용하고 있는데 일관된 형식을 따르도록 수정하는게 어떨까요?

Ra99it replied:
적극 반영했습니다.

f-lab-johnp commented on LoL_PlayerEventGenertor/src/main/java/player/Room.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    CountDownLatch latch = new CountDownLatch(player);
+    ExecutorService executor = Executors.newFixedThreadPool(player);
```
앞서 Main에서 쓰레드 풀을 생성한 다음, 내부에서 다시 새로운 쓰레드 풀을 만들고 있는 것으로 보입니다.
위와 같은 구조에서는 자원이 낭비될 수 있고 적절히 관리되지 않으면 메모리 누수나 데드락이 발생할 수 있을 것 같아요.
상위에서 생성된 쓰레드 풀을 재사용하는건 어떨까요?

```suggestion
    private final ExecutorService executorService;

    public Room(..., ExecutorService executorService) {
        ...
        this.executorService = executorService;
    }
    CountDownLatch latch = new CountDownLatch(player);
```

Ra99it replied:
Main에서 Room 쓰레드 풀을 생성하고, Room에서 10개의 Player 쓰레드 풀을 생성합니다.
1개의 Room에는 10명의 플레이어가 존재합니다.
- Room
  - Player
  - Player
  -   ....

Main에 있는 Room 쓰레드 풀을 재사용하게 되면 10개의 플레이어 쓰레드 풀을 생성하지 못하는 거 같습니다!

f-lab-johnp replied:
Room에서 쓰레드 풀 사이즈를 늘려서 사용하는 방법을 생각했는데 격리해야한다면 지금처럼 유지해도 괜찮을 것 같네요.
보통 작업의 성격이나 수명주기가 다른 경우에 격리하기도 합니다 👍🏻️

Ra99it committed:
style, docs: google style guide 적용 및 함수 javadoc 문서화

- 코드 가독성을 위한 Code Convention : Google style guide 적용\n- 설명이 필요한 함수 javadoc 문서화 진행

Ra99it committed:
Merge pull request #8 from f-lab-edu/feature/#4

feat : Docker Swarm 환경 Trino Cluster 구축

Ra99it committed:
refactor: 메서드 이름 변경 및 코드 간결화를 진행

메서드의 이름을 좀 더 직관적으로 알 수 있게 변경하였습니다. 또한 챔피언을 생성하는 메서드인 getChampions()의 소스를 간결하게 변경하였습니다.

Ra99it committed:
Merge branch 'develop' of https://github.com/f-lab-edu/league-of-legends-data-solution into develop

Ra99it committed:
refactor: Kafka Produce의 호스트 정보를 외부로 받게 했습니다.

소스 코드에 있었던 호스트 정보를 보안상 외부에서 입력하도록 변경하였습니다.

Ra99it committed:
fix: args 버그 수정

조건문을 args.length > 0으로 잘못된 코드로 프로그램이 실행되지 않아서 수정

Ra99it committed:
chore: 사용하지 않거나 불필요한 파일 삭제

Ra99it committed:
chore: 불필요한 파일 삭제

Ra99it committed:
Merge pull request #7 from f-lab-edu/feature/#3

feat : Docker Swarm 환경 Kafka Cluster 추가

Ra99it committed:
Merge pull request #9 from f-lab-edu/feature/#5

feat : Docker Swarm 환경 Airflow 환경 구축

Ra99it committed:
refactor: 테스트를 위한 sysout 삭제

Ra99it committed:
feat: 지정한 경로에 로그를 남길 수 있는 기능을 추가했습니다.

원래의 경우 Kafka Producer를 통해 직접 연결하는 소스만 존재했는데, 지정한 경로의 json형식으로 남길 수 있도록 소스를 변경하였습니다.

f-lab-johnp commented:
이후부터는 develop → main PR은 운영 릴리즈로 생각하시면 될거 같아요.
이제 앞서 feature → develop 단계에서 리뷰를 완료했기 때문에 릴리즈 단계에서는 코드리뷰보다는 배포 공유 정도로 보시면 됩니다.
수고 많으셨습니다!

Ra99it committed:
refactor: 로그를 안정적으로 남길 수 있도록 대폭 개선

logback을 활용해서 조금 더 안정적이고, 재사용하기 용이하게 변경하였습니다.

Ra99it committed:
feat: logback 의존성 추가

Ra99it committed:
chore: 보안을 위해 Player.Log 삭제

Ra99it committed:
Merge pull request #12 from f-lab-edu/feature/#10

feat: 지정한 경로에 로그를 남길 수 있는 기능을 추가했습니다.
